File: rewrite-ruby/src/test/java/org/openrewrite/ruby/internal/StringUtilsTest.java
/*
 * Copyright 2023 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.ruby.internal;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;

class StringUtilsTest {

    @ParameterizedTest
    @ValueSource(strings = {"%[", "%s[", "%q[", "%Q["})
    void percentDelimiters(String delimiter) {
        assertThat(StringUtils.endDelimiter(delimiter)).isEqualTo("]");
    }

    @ParameterizedTest
    @ValueSource(strings = {"\"", "'", "`"})
    void stringDelimiters(String delimiter) {
        assertThat(StringUtils.endDelimiter(delimiter)).isEqualTo(delimiter);
    }

    @Test
    void empty() {
        assertThat(StringUtils.endDelimiter("")).isEqualTo("");
    }
}

File: rewrite-core/src/test/kotlin/com/netflix/rewrite/internal/StringUtilsTest.kt
/*
 * Copyright 2020 the original authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.internal

import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.assertEquals
import org.junit.Test

class StringUtilsTest {
    @Test
    fun detectIndentLevel() {
        assertThat(StringUtils.indentLevel("""
            |<
            |   <
            |   <
            |   <
            |<
        """.trimMargin())).isEqualTo(0)

        assertThat(StringUtils.indentLevel("""
            |<
            |   <
        """.trimMargin())).isEqualTo(3)

        assertThat(StringUtils.indentLevel("""
            |<
            |    <
            |      <
        """.trimMargin())).isEqualTo(4)

        // ignores the last line if it is all blank
        assertThat(StringUtils.indentLevel("""
            class {
                A field;
            }
        """)).isEqualTo(12)

        assertThat(StringUtils.indentLevel("""
            | <
            |  <
            |   <
            |    <
        """.trimMargin())).isEqualTo(1)

        assertThat(StringUtils.indentLevel("""
            | <
            |  <
            |    <
            |    <
            |      <
        """.trimMargin())).isEqualTo(1)

        assertThat(StringUtils.indentLevel("""
            |<
            |<
        """.trimMargin())).isEqualTo(0)

        // doesn't consider newlines that occur as the first character on the first line or terminating newlines
        assertThat(StringUtils.indentLevel("""
            |
            |  <
            |    <
            |    <
            |
        """.trimMargin())).isEqualTo(2)
    }

    @Test
    fun trimIndent() {
        val input = """
            class {
                A field;
            }
        """

        assertEquals(input.trimIndent(), StringUtils.trimIndent(input))
    }

    @Test
    fun trimIndentMinimalIndent() {
        val input = """
        class {
                A field;
            }
        """

        assertEquals(input.trimIndent(), StringUtils.trimIndent(input))
    }

    @Test
    fun trimIndentNoIndent() {
        val input = "class{\n   A field;\n}"

        assertEquals(input.trimIndent(), StringUtils.trimIndent(input))
    }
}